soteria/lib/symex/tree_block.ml

```diff
@@ -0,0 +1,674 @@
+      let++ _, tree = frame_range st ~replace_node ~rebuild_parent range in
+      tree
+
+    let produce (serialized : MemVal.serialized) (range : Range.t) (st : t) :
+        t Symex.t =
+      let replace_node t =
+        let+ t = MemVal.produce serialized t in
+        Ok t
+      in
+      let rebuild_parent = of_children in
+      let* res = frame_range st ~replace_node ~rebuild_parent range in
+      match res with Ok (_, tree) -> return tree | _ -> Symex.vanish ()
+  end
+
+  type t = {
+    root : Tree.t;
+    bound : sint option; [@printer Fmt.(option ~none:(any "⊥") Symex.Value.ppa)]
+  }
+  [@@deriving show { with_path = false }]
+
+  let is_empty t = Option.is_none t.bound && Tree.is_empty t.root
+
+  let pp_pretty ft t =
+    let open PrintBox in
+    let r = ref [] in
+    let () =
+      Option.iter
+        (fun b ->
+          let range_str = Fmt.str "[%a; ∞[" Symex.Value.ppa b in
+          r := [ (range_str, text "OOB") ])
+        t.bound
+    in
+    let () =
+      Tree.iter_leaves_rev t.root (fun leaf ->
+          let range_str = (Fmt.to_to_string Range.pp) leaf.range in
+          let node_str = Fmt.to_to_string Node.pp leaf.node |> text in
+          r := (range_str, node_str) :: !r)
+    in
+    PrintBox_text.pp ft (frame @@ record !r)
+
+  (** Logic *)
+
+  type serialized_atom =
+    | MemVal of {
+        offset : sint; [@printer Symex.Value.ppa]
+        len : sint; [@printer Symex.Value.ppa]
+        v : MemVal.serialized;
+      }
+    | Bound of sint [@printer fun f v -> Fmt.pf f "Bound(%a)" Symex.Value.ppa v]
+  [@@deriving show { with_path = false }]
+
+  type serialized = serialized_atom list
+
+  let iter_values_serialized serialized f =
+    List.iter (function MemVal { v; _ } -> f v | _ -> ()) serialized
+
+  let of_opt ?(mk_fixes = fun () -> Symex.return []) = function
+    | None ->
+        let+ fixes = mk_fixes () in
+        Missing fixes
+    | Some t -> Result.ok t
+
+  let to_opt t = if is_empty t then None else Some t
+
+  let with_bound_check (t : t) (ofs : sint) f =
+    let** () =
+      match t.bound with
+      | None -> Result.ok ()
+      | Some bound ->
+          if%sat bound <@ ofs then Result.error `OutOfBounds else Result.ok ()
+    in
+    let++ res, root = f t.root in
+    (res, to_opt { t with root })
+
+  let with_bound_and_owned_check ?mk_fixes t ofs f =
+    let** t = of_opt ?mk_fixes t in
+    with_bound_check t ofs f
+
+  let assert_exclusively_owned t =
+    let** t = of_opt t in
+    match t.bound with
+    | None -> miss_no_fix ~msg:"assert_exclusively_owned - no bound" ()
+    | Some bound ->
+        let { range = low, high; node; _ } = t.root in
+        if Node.is_fully_owned node then
+          if%sat low ==@ zero &&@ (high ==@ bound) then Result.ok ()
+          else
+            miss_no_fix
+              ~msg:"assert_exclusively_owned - tree does not span [0; bound[" ()
+        else
+          miss_no_fix ~msg:"assert_exclusively_owned - tree not fully owned" ()
```

Comment:
This is wrong in the context of MemVal potentially being partially-owned.

How about something of the form `Node.assert_exclusively_owned` that itself calls `MemVal.assert_exclusively_owned`?